### DISCUSSION TITLE
Path comparision with Wald Test

### DISCUSSION THREAD

**Discussion by PascalHessler** [sem-in-r/seminr#357] [Ideas]
Hey, 
so I have a case where we are interested if one path has a higher effect than another one. In "Normal" regression, you can just use the Wald test for this. Technically this should be possible for pls-sem to (Both paths are to one construct). 
I would like to have the option to use the Wald test, but I was not sure if this is an "issue" of this package, so I am posting it here. What do you think?
Best regards

**Comment by soumyaray** [comment]
Dear @PascalHessler thank you for your excellent enquiry.  I have not used the Wald test in practice and am mostly familiar with it as a test of an individual coefficient, or a set of coefficients, being significantly different from zero. It would be great if you can share a resource of the Wald test being use to compare the relative level of two coefficients. Searching casually, I also see people offering [t-tests for comparing coefficients](https://stats.stackexchange.com/questions/93540/testing-equality-of-coefficients-from-two-different-regressions).  If there is well vetted and accepted method available in practice, we might consider incorporating it into SEMinR – though we are quite conservative about adding new methods as many prove to be problematic over the course of time.

In principle, any test that only needs to know the estimated coefficients and their respective standard errors is implementable. The regular PLS estimation can provide the coefficients and bootstrapping can provide their standard errors. If you have a formula for a relevant test, you (or we) can readily compute it with these.  Again, if you can provide a reference that offers a clear cut description of the procedure you are interested in, we can at least work with you to put together an ad hoc function that you can use – and later debate if it should be formally included as a publicly exported function of SEMinR.

**Reply by PascalHessler** [reply]
Dear @soumyaray, thank you for your offer to help and the fast answer. I completely understand that implementing many tests does bring a lot of issues. But I think this is not necessary.

My main issue was that I did not know how to access all the information I needed. For example, I did not know how to access the variance-covariance matrix of the relevant part in the sem. (From my understanding (please correct me if I am wrong), the PLS-SEM calculates for the inner model basically regular regressions for each part of it. When I want to compare two paths of such a regression, this should work exactly as if we would not have used sem, so i just need to access this regression)

## Examples
I did find two packages that implement the test:
The `aod` library
simple example: https://www.statology.org/wald-test-in-r/
doc: https://www.rdocumentation.org/packages/aod/versions/1.3.3/topics/wald.test
and the `ClubSandwich `
Advanced but with many use cases: https://cran.r-project.org/web/packages/clubSandwich/vignettes/Wald-tests-in-clubSandwich.html 
Doc: https://www.rdocumentation.org/packages/clubSandwich/versions/0.5.11/topics/Wald_test
Also, I do not need the clustering stuff

As you see, they both basically only need the variance-covariance matrix coefficients and which terms you want to compare. If you can show me how I can access that information, I think I can do the rest myself. Perhaps this is actually easy to do, but I am not used to R (normally programming in python) and did not find an easy solution.

### Stata
Here is a forum post on how to use the wald-test in Stata: https://www.statalist.org/forums/forum/general-stata-discussion/general/1311208-wald-test 

A small example also from stata:
```
sureg (mediator explain) (dv mediator explain)
test mediator=explain
```
Unfortunately, Stata does not make their code available....

**Reply by soumyaray** [reply]
Dear @PascalHessler, unfortunately with PLS models we do not operate on a variance-covariance matrix alone, and producing one of the constructs is possible by doing `cov(pls_model$construct_scores)`, but is misleading because it does not account for measurement errors. What you can get is the coefficients of the structural model and their standard errors derived from a bootstrap. But as far as I know, that might only help you conduct the Wald test on a single coefficient rather than on bundles of coefficients (or all the coefficients). Unless of course you find a package or formula that operate only on coefficients and their standard errors. Would that be of any help?

**Reply by soumyaray** [reply]
@PascalHessler now that I think about it, I think we can get the variance-covariance matrix of coefficients (if that is what Wald test requires) from the bootstrapped coefficients. If your bootstrapped pls model is called `boot_pls`, then `boot_pls$boot_paths` gives a list of path matrices for your `b` bootstrap samples. However, you would have to convert that list into a matrix of coefficients (columns) versus replicates (rows). And then you can simply run `cov()` on that matrix. If this makes sense and is important to you, I can try to work with you to do that.

**Reply by PascalHessler** [reply]
@soumyaray Ok, I am trying to understand what you are suggesting. Here is my working example:
``` R
library(seminr)
corp_rep_data <- seminr::corp_rep_data
corp_rep_mm <- constructs(
  composite("COMP", multi_items("comp_", 1:3)),
  composite("LIKE", multi_items("like_", 1:3)),
  composite("CUSA", single_item("cusa")),
  composite("CUSL", multi_items("cusl_", 1:3))
  )


corp_rep_sm <- relationships(
  paths(from = c("COMP", "LIKE"), to = c("CUSA", "CUSL")),
  paths(from = c("CUSA"), to = c("CUSL")))

pls_model <- estimate_pls(data = corp_rep_data,
                                   measurement_model = corp_rep_mm,
                                   structural_model  = corp_rep_sm,
                                   inner_weights = path_weighting,
                                   missing = mean_replacement,
                                   missing_value = "-99")

boot_pls <- bootstrap_model(seminr_model = pls_model,
                                 nboot = 1000,
                                 cores = NULL,
                                 seed = 123)
```
When i have the boot_pls$boot_paths i get a table like this times n bootstrap:
`boot_pls$boot_paths[, , 1]`
| COMP | LIKE  |    CUSA|  CUSL| 
| -------- | ----- | ------ |------ |
| COMP |   0   | 0 0.1445186 |0.02511431| 
| LIKE   | 0   | 0 0.5040798 |0.32057445| 
| CUSA |   0  |  0 0.0000000 |0.50875271| 
| CUSL |   0   | 0 0.0000000 |0.00000000| 

And the goal is to get a table like this?
| index        | COMP          | LIKE| CUSA |
| -------- | ----- | ------ |------ |
| 0 Boot step  | 0.02511431   | 0.32057445 | 0.50875271|
| 1 Boot step  | Path cof on y   | Path cof on y  | Path cof on y|

In this case, I only would select the last column of the first table, but I am not quite sure about this. Can you confirm this?
This would be the programming code (4 is in this case the last column)
`cov(t(apply(boot_pls$boot_paths, 3, function(x) x[, 4])))`
Output: 
|       | COMP         | LIKE         | CUSA         | CUSL |
|-------|--------------|--------------|--------------|------|
| COMP  | 0.003097094  | -0.0020995256| -0.0004882490| 0    |
| LIKE  | -0.002099526 | 0.0031877469 | -0.0009827908| 0    |
| CUSA  | -0.000488249 | -0.0009827908| 0.0017738661 | 0    |
| CUSL  | 0.000000000  | 0.0000000000 | 0.0000000000 | 0    |

Just tested it with the wald test from the aod pckage:
``` R
library(aod)
cov_test <- cov(t(apply(boot_pls$boot_paths, 3, function(x) x[, 4])))
coef <- boot_pls$path_coef[, ncol(boot_pls$path_coef)]
wald.test( Sigma=cov_test, b= coef, Terms =1:3 )
```
But honestly, I'm not quite sure if it makes sense what I am doing here. 

Best and a huge thank you for helping me out!

**Reply by soumyaray** [reply]
Hi @PascalHessler, in theory, what you are doing is what I was suggesting.

I've developed a generalizable (but experimental) `vcov` and `coef` type functions for bootstrapped seminr models that you can find at: https://github.com/sem-in-r/seminr/blob/ray/experiment-wald-test/R/wald.R

At the bottom of that file, you'll find a commented-out section with two examples:
- a Wald test of a linear model using single measurement items from the mobi dataset (exogenous item is Loyalty item `CUSL1`)
- a Wald test of the PLS model of the corresponding constructs (exogenous construct is Loyalty, renamed to `LOY`)

Feel free to uncomment it execute it, play with some variations, and see if it seems reasonable.

I cannot pretend that this is a rigorous and vetted approach for PLS models – I have not seen it before and we're extrapolating from what we have seen of other types models. Moreover, one must limit this application to a single endogenous variable, as the coefficients must only be of its exogenous variables. Up to that limit, I think this might be naïvely defensible to some degree!

Would love to get some feedback, especially if you share it with others who know more about the Wald test.

**Reply by PascalHessler** [reply]
Hi @soumyaray,

This goes beyond what I expected—honestly, this has been my best experience interacting with a package developer. Thank you for your help and dedication—please keep up the great work!

I’ll definitely reach out again once I gather some feedback. Unfortunately, I don’t currently know anyone with deep expertise in the Wald test, but if I do meet someone, I’ll be sure to share it with you.

**Reply by soumyaray** [reply]
Hi @PascalHessler thanks for starting the conversation and providing the starter code – it made writing up the rest quite easy. I think you can even use the tentative solution in a paper if you wish, if you describe the overall approach in a footnote or online appendix: obtaining bootstrapped coefficients to compute their variance-covariance matrix, as needed for input into the Wald test.

**Reply by PascalHessler** [reply]
Hi, @soumyaray, 

so i implemented the test for my model. After some smaller issues I achieved my goal :) Perhaps you can add this in the commented out section in the package? 

```
# b <- coef.boot_seminr_model(boot_obj = boot_pls, endogenous = "LOY")
# S <- vcov.boot_seminr_model(boot_obj = boot_pls, endogenous = "LOY")
# Tests the joined hypotheses of 'SAT->LOY' == 0 and 'COM->LOY' == 0
# Corresponding L Matrix = [0 1 0, 0 0 1]
# wald.test(b=b, Sigma=S, Terms=c(2,3))

# If you want to test if to pathes are equal to each other you need to provide the L argument representing the linear combinations 
# of the coefficients to be tested
# For example: 'SAT->LOY'=='COM->LOY'  would correspond to a Matrix [0  1 -1]
# wald.test(b=b, Sigma=S, L= matrix(c(0,1,-1), nrow = 1))
```

Perhaps this helps someone else in the future :)

**Reply by soumyaray** [reply]
Thanks so much @PascalHessler – could you link me to any resources discussing the `L` matrix and `nrows` parameter you used in more detail? The help file for `wald.test` is very brief on them and doesn't help me understand your example quite yet.

**Reply by PascalHessler** [reply]
Hey @soumyaray  it is based on the article shared above: https://cran.r-project.org/web/packages/clubSandwich/vignettes/Wald-tests-in-clubSandwich.html.
Basically 0 1 - 1 Represents "Sat->Loy" - "Com->Loy " = 0. A standard t-test for only Sat->Loy = 0 would be represented by 0 1 0

Hope this helps.

the `nrows `parameter is not from the package just setting the dimensions of the matrix to consist out of one row. If not R would create one column and 3 rows

**Reply by soumyaray** [reply]
perfect – i'll give this a read when i have a moment.  It seems to be that setting the values of L could be the kind of thing for which we could provide a high level function that makes it easier.

**Reply by PascalHessler** [reply]
Yes that would be much easier. The clubsandwich provides such functions but they are not compatible, since they not only return the matrix but add addintional information.
